xdsl/dialects/vector.py

```diff
@@ -36,10 +30,13 @@ def verify_(self):
     @staticmethod
     def get(ref: SSAValue | Operation,
             indices: List[SSAValue | Operation]) -> Load:
+        ref = SSAValue.get(ref)
+        assert isa(ref.typ, MemRefType[Attribute])
```

Comment:
I don't quite like the assert here. Above, it seems that the asserts are on things that irdl should check already, but this one in particular works on user input. I would therefore expect a better error message, instead of just an assertion failure in this case.

Reply:
That's an interesting point. There's a few ways of looking at it, and one is user error vs programmer error, which in this case is a bit weird given that we're writing a compiler framework as opposed to an app. In a previous life of an Objective-C developer I strongly advocated for asserting things that I was hoping the compiler would enforce, like in this case. Since this code isn't on a path of user input processing, which is mostly when parsing files, it seems ok to assert, to explode as soon as you've found that something has gone wrong. This same property will be checked in a pretty way in the `validate` function, but it feels ok to be less pretty about the validation when a user of the API passed a value of the wrong type.

Reply:
I think we are once again running into the issue that we are building two things in the repo. From the standpoint of core xDSL (and that is my standpoint) the API of the dialects, in particular constructors, are user-facing API and, hence, need good error reports IMO.

Looking from the standpoint of someone who just wants to use xdsl-opt to optimize their code , these constructors will only be used in rewriting, which will be provided by us, and hence this is not public API...

Reply:
Pushed a version with a pretty assertion error, what do you think about this format?

```
AssertionError: Expected value of type xdsl.dialects.memref.MemRefType[xdsl.ir.Attribute], got value of type IntegerType
```

Reply:
Yeah, I like it :+1:

Reply:
Actually, we shouldn't use an assert here, now that I think of it.
We should instead use `raise ValueError`, which convey a better meaning.

`assert`, in my opinion, are meant to me triggered we have a bug in `xDSL`. However, exceptions are meant to be thrown when the user made a mistake.

What do you think about this?

Reply:
Weak agree from me.

Reply:
Yeah makes sense